app/models/backend_api_config.rb

```diff
@@ -21,6 +21,8 @@ class BackendApiConfig < ApplicationRecord
   validates :path, uniqueness: { scope: :service_id, case_sensitive: false, message: "This path is already taken. Specify a different path." }
   validates :path, length: { in: 1..255, allow_nil: false }, path: true
 
+  before_destroy :destroy_usage_limits
```

Comment:
```suggestion
    has_many :usage_limits, ->(backend_usage) { of_metric(backend_usage.backend_api_metrics) }, through: :service, dependent: :destroy
```

When querying the result is:
```sql
[1] pry(main)> BackendApiConfig.take.usage_limits
  BackendApiConfig Load (0.6ms)  SELECT `backend_api_configs`.* FROM `backend_api_configs` ORDER BY `backend_api_configs`.`id` ASC LIMIT 1
  UsageLimit Load (4.3ms)  SELECT `usage_limits`.* FROM `usage_limits` INNER JOIN `plans` ON `usage_limits`.`plan_id` = `plans`.`id` INNER JOIN `services` ON `plans`.`issuer_id` = `services`.`id` WHERE `plans`.`type` = 'ApplicationPlan' AND `services`.`id` = 1 AND `usage_limits`.`plan_type` = 'Plan' AND `plans`.`issuer_type` = 'Service' AND `usage_limits`.`metric_id` = NULL
```

Reply:
Actually even better:
```suggestion
  has_many :usage_limits, ->(backend_usage) { of_metric(backend_usage.backend_api_metrics.all) }, through: :service, dependent: :destroy
```

Then a single query is being run:
```sql
[1] pry(main)> BackendApiConfig.take.usage_limits
  BackendApiConfig Load (0.4ms)  SELECT `backend_api_configs`.* FROM `backend_api_configs` ORDER BY `backend_api_configs`.`id` ASC LIMIT 1
  UsageLimit Load (1.7ms)  SELECT `usage_limits`.* FROM `usage_limits` INNER JOIN `plans` ON `usage_limits`.`plan_id` = `plans`.`id` INNER JOIN `services` ON `plans`.`issuer_id` = `services`.`id` WHERE `plans`.`type` = 'ApplicationPlan' AND `services`.`id` = 1 AND `usage_limits`.`plan_type` = 'Plan' AND `plans`.`issuer_type` = 'Service' AND `usage_limits`.`metric_id` IN (SELECT `metrics`.`id` FROM `metrics` INNER JOIN `backend_apis` ON `metrics`.`owner_id` = `backend_apis`.`id` WHERE `backend_apis`.`id` = 1 AND `metrics`.`owner_type` = 'BackendApi' ORDER BY `metrics`.`id` ASC)
```

Reply:
This should reduce number of queries and delegate the dark magic to rails.

Reply:
Great job @akostadinov constructing those associations! It's very interesting.

The last association is nice because of its efficiency.

However... I am quite reluctant to change the current implementation to what you are proposing, because I think that it is still a bit confusing...

There is this `through: :service` here, and on the Service model it's `through: :application_plans`, so I think it's a bit too deep.

But anyway, I decided to try, and while the query itself works perfectly, when I try to delete a BackendApiConfig, I am getting 
![Screenshot from 2025-02-19 22-42-01](https://github.com/user-attachments/assets/78abcfdb-11bb-458d-8e49-29157945efe1)

https://github.com/rails/rails/blob/v7.0.8.7/activerecord/lib/active_record/associations/has_many_through_association.rb#L131
https://github.com/rails/rails/blob/v7.0.8.7/activerecord/lib/active_record/associations/through_association.rb#L104

I think this is so by design. I am not sure whether this has some workaround, but I personally wouldn't over-complicate the things, while we have a working solution (and I don't think it's too bad).

Reply:
It's a bummer because it doesn't really make sense to me. Adding records fine, but what's their issue with destroy? 

Anyway, then it makes sense to keep the association with `dependent` not specified or if needed `dependent: false` and then in your `before_destroy` hook, do `usage_limits.destroy_all` or if this still doesn't work, something like
```rb
usage_limits.find_each(&:destroy!)
```

This would at least provide a convenience to access these usage limit objects for other purposes as well.

Reply:
Or we can schedule deletion in the background. Idk whether we can have many usage limits to make it worth.

Reply:
Probably the nested association or the background deletion would work as well, but I don't think those approaches are obviously better than the callback, just another way to do the same. If it's already done through a callback, I think that's fine.

Also, would it be _conceptually_ correct to add an association with the sole purpose of auto-deleting? not being used for anything else? I can see myself in the future forgetting about this, finding the association by accident, checking nobody calls it, and removing it.

Reply:
@akostadinov `usage_limits.destroy_all` has the same issue, but `usage_limits.find_each(&:destroy!)` works.


Also, I just realized that in your query `:of_metric` scope (which is `where(:metric_id => metric)`) is used with a relation ("array" of `Metric` instances), and ActiveRecord is still able to resolve it! I previously though, it would be either a list of IDs or a single `Metric` instance. Crazy.

Reply:
There is one more thing which might have a long term meaning. When deleting we may set `destroyed_by_association`. e.g.

```rb
usage_limits.find_each do | usage_limit |
  usage_limit.destroyed_by_association = "dummy_association_may_figure_out_proper_value_later"
  usage_limit.destroy!
end
```

Idk if it is worth it. Just thought to speak about it now I'm into this kind of stuff. Some of our callbacks check whether we destroy by association or not.

Reply:
Actually perhaps it doesn't make sense to set `destroyed_by_association` because it seems to usually be used for protective purposes of the primary associated object. Or to skip steps like notifications and such. So in this case we are doing a standalone full removal. That's why I would keep it as is. Already approved the whole thing. Very nice investigation!